torch/distributed/_shard/checkpoint/state_dict_saver.py

```diff
@@ -123,32 +129,76 @@ def save_state_dict(
         >>>     storage_writer=fs_stroage_writer,
         >>> )
     """
-    (
-        metadata,
-        bytes_write_requests,
-        tensor_write_requests,
-    ) = _prepare(state_dict)
-
-    is_rank0 = not dist.is_initialized() or dist.get_rank(process_group) == 0
-    if is_rank0:
-        storage_writer.prepare()
+    is_coordinator = not dist.is_initialized() or dist.get_rank(process_group) == coordinator_rank
+    data: List[Optional[BaseException]] = [None]
+    if is_coordinator:
+        try:
+            storage_writer.prepare()
+        except BaseException as e:
+            data = [e]
 
     # Writing can only start once prepare has finished
     if dist.is_initialized():
-        dist.barrier(process_group)
-
-    combined_writes: List[Union[TensorWriteRequest, BytesWriteRequest]] = []
-    combined_writes.extend(tensor_write_requests)
-    combined_writes.extend(bytes_write_requests)
+        dist.broadcast_object_list(data, group=process_group, src=coordinator_rank)
+
+    if data[0] is not None:
```

Comment:
nit, if this is actually the exceptioons, shall we name this `exceptions` instead?

Reply:
good point, `data` is a terrible name here.